### ISSUE TITLE
Using on esp32 with embassy hangs on `Epd2in9bc::new`

### ISSUE THREAD

**Issue by JohnTheCoolingFan** [caemor/epd-waveshare#220]
Hello, I am trying to drive an epaper display (2.9inch b/w/r from weact, analogous to waveshare module, working with GxEPD2) using this library, running on esp32 using embassy. The lack of examples really makes it difficult, but I'm trying to figure it out.
One problem is that the library doesn't allow to just use the `esp_hal::spi::master::Spi` and instead one has to "wrap" it through `embassy_embedded_hal::blocking::spi::SpiDevice`, which itself requires a lot of setup.

So, here's my code so far, based on a template from [`esp-generate`](https://github.com/esp-rs/esp-generate):
```rs
#![no_std]
#![no_main]

use core::cell::RefCell;

use embassy_embedded_hal::shared_bus::blocking::spi::SpiDevice;
use embassy_executor::Spawner;
use embassy_sync::blocking_mutex::NoopMutex;
use embassy_time::{Duration, Timer};
use embedded_graphics::{
    prelude::{Point, Primitive},
    primitives::{Line, PrimitiveStyle},
    Drawable,
};
use esp_backtrace as _;
use esp_hal::{
    gpio::{Input, Level, NoPin, Output, Pull},
    prelude::*,
    spi::{
        master::{Config, Spi},
        SpiMode,
    },
    Blocking,
};
use log::info;

extern crate alloc;

use epd_waveshare::{epd2in9bc::*, prelude::*};
use static_cell::StaticCell;
#[main]
async fn main(spawner: Spawner) {
    let peripherals = esp_hal::init({
        let mut config = esp_hal::Config::default();
        config.cpu_clock = CpuClock::max();
        config
    });

    esp_alloc::heap_allocator!(72 * 1024);

    esp_println::logger::init_logger_from_env();

    let timer0 = esp_hal::timer::timg::TimerGroup::new(peripherals.TIMG1);
    esp_hal_embassy::init(timer0.timer0);

    info!("Embassy initialized!");

    /*
    let timer1 = esp_hal::timer::timg::TimerGroup::new(peripherals.TIMG0);
    let _init = esp_wifi::init(
        timer1.timer0,
        esp_hal::rng::Rng::new(peripherals.RNG),
        peripherals.RADIO_CLK,
    )
    .unwrap();
    */

    info!("Initializing spi bus");

    static SPI_BUS: StaticCell<NoopMutex<RefCell<Spi<'static, Blocking>>>> = StaticCell::new();
    let spi_bus = Spi::new(peripherals.SPI2)
        //.with_cs(peripherals.GPIO5)
        .with_cs(NoPin)
        .with_sck(peripherals.GPIO18)
        .with_mosi(peripherals.GPIO23)
        .with_miso(NoPin);
    let spi_bus = NoopMutex::new(RefCell::new(spi_bus));
    let spi_bus = SPI_BUS.init(spi_bus);

    info!("Initializing spi device");

    let mut spi = SpiDevice::new(spi_bus, Output::new(peripherals.GPIO5, Level::Low));

    info!("Initializing pins");

    let busy_in = Input::new(peripherals.GPIO4, Pull::Down);
    let dc = Output::new(peripherals.GPIO17, Level::Low);
    let rst = Output::new(peripherals.GPIO16, Level::High);

    let mut delay = embassy_time::Delay;

    info!("Initializing epd");

    let mut epd =
        Epd2in9bc::new(&mut spi, busy_in, dc, rst, &mut delay, None).expect("EPD creation error");

    info!("Drawing");

    let mut mono_display = Display2in9bc::default();

    let _ = Line::new(Point::new(0, 120), Point::new(0, 200))
        .into_styled(PrimitiveStyle::with_stroke(Color::Black, 1))
        .draw(&mut mono_display);

    let mut chromatic_display = Display2in9bc::default();

    let _ = Line::new(Point::new(15, 120), Point::new(15, 200))
        .into_styled(PrimitiveStyle::with_stroke(Color::Black, 1))
        .draw(&mut chromatic_display);

    info!("Updating color frame");

    epd.update_color_frame(
        &mut spi,
        &mut delay,
        mono_display.buffer(),
        chromatic_display.buffer(),
    )
    .unwrap();

    epd.display_frame(&mut spi, &mut delay).unwrap();

    info!("Pre sleep");

    epd.sleep(&mut spi, &mut delay).unwrap();

    info!("Post sleep");

    // TODO: Spawn some tasks
    let _ = spawner;

    loop {
        info!("Hello world!");
        Timer::after(Duration::from_secs(1)).await;
    }

    // for inspiration have a look at the examples at https://github.com/esp-rs/esp-hal/tree/v0.22.0/examples/src/bin
}
```

With the debug logs in place, the point in the program where it stops is after `info!("Initializing epd");`.
Did I do something obviously wrong in my code? What can I do to diagnose the issue further to find out why it's hung?

**Comment by JohnTheCoolingFan** [comment]
Potentially related issues, none of which have an answer:
- #140 faulty original code and the comment focuses on the code part that was added to test the spi and just gives a vague example with their HAL
- #155 No answer from original author
- #70 old and outdated, uses a different module, code links are not permalinks so no idea what was the library code linked

**Comment by JohnTheCoolingFan** [comment]
Also, a pin mapping:
| EPD | ESP32 |
|----------|-------|
| GND | GND |
| VCC | 3V3 |
| SDA | P23 (VSPI MOSI) |
| SCL | P18 (VSPI SCLK) |
| CS | P5 |
| D/C | P17 |
| RES | P16 |
| BUSY | P4 |

**Comment by JohnTheCoolingFan** [comment]
I've checked some things and I think this library has a problem with `Epd2in9bc::new` or `Epd2in9bc::init` specifically.

I've used https://github.com/GnomedDev/epaper-simplyplural-badge/tree/0440c798289a3b36dcfb42f4938d9fc3a9111da9 as a base project to verify whether the project that was linked to me would work if I tweak it to my hardware. I've done some incremental changes: first changing gpio pins, then adding logging, and at the end changing from `epd2in13_v2` to `epd2in9bc`. The last one caused the program to break. Source code here: https://github.com/JohnTheCoolingFan/epaper-simplyplural-badge/tree/f58ac26fcb97d9d7a89a55d27dc0b1b0afd37201
